libbeat/api/routes.go

```diff
@@ -38,6 +40,10 @@ func NewWithDefaultRoutes(log *logp.Logger, config *common.Config, ns lookupFunc
 	mux.HandleFunc("/state", makeAPIHandler(ns("state")))
 	mux.HandleFunc("/stats", makeAPIHandler(ns("stats")))
 	mux.HandleFunc("/dataset", makeAPIHandler(ns("dataset")))
+
+	for api, h := range handlerFuncMap {
+		mux.HandleFunc(api, h)
+	}
```

Comment:
It would be nice if we can dynamically add/remove routes. E.g. when starting/removing sub-systems. The problem with a "global" like this one is initialization order. You can not tell when `NewWithDefaultRoutes` is called. This requires anyone to call `AddHandlerFunc` from within a `init` function. The response for routes can be changed (via globals), but the presence/behavior can hardly be configured.

Reply:
Thanks @urso ! it makes sense for adding/removing routes. I had an implementation for this, please have a check.

Reply:
There is still a "global" routes in the implementation that shares its `http.ServeMux` with the server initialized in `NewWithDefaultRoutes`. The reason for this is that I think there needs to be a "global" thing, either routes or a "global" `api.Server`, in order to allow register/deregister routes, and though there are other ways to initialize `api.Server` such as its `New` function, but I don't see it's being used except `NewWithDefaultRoutes`.